Allow `UniqueConstraint` in place of `unique_together` for `TranslatableMixin`'s system check

Temidayo32 opened the pull request:
Django `unique_together` may be deprecated in the future. `models.UniqueConstraint` is recommended [in Django documentation as a good replacement as it also offer more functionality](https://docs.djangoproject.com/en/4.2/ref/models/options/#unique-together:~:text=UniqueConstraint%20provides%20more%20functionality%20than%20unique_together.%20unique_together%20may%20be%20deprecated%20in%20the%20future.). This pull request resolves this issue #11098







_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

Temidayo32 commented:
I would need to modify some tests so the checks can run successfully.

Temidayo32 commented:
@lb- I have a `test__migrations (wagtail.tests.test_migrations.TestForMigrations)` failure. Not sure what to do. I can't find `manage.py` in `wagtail.tests`, so I am not sure whether create one to create the migration files for this test to be successful.

this is the specific error I am getting:

> 


Traceback (most recent call last):
  File "/home/danajayi/wagtail/wagtail/tests/test_migrations.py", line 63, in test__migrations
    self.fail("Model changes with no migrations detected:\n%s" % migrations)

lb- commented on docs/extending/extending_draftail.md:
````diff
@@ -422,6 +436,7 @@ Here is an example with highlighting of problematic punctuation – first, regis
 
 ```python
 from wagtail.admin.rich_text.editors.draftail.features import DecoratorFeature
+from wagtail import hooks
````
I recommend you just add these docs improvements as a separate PR. Will be easy to merge and avoids confusion on this PR.

Temidayo32 replied:
Yeah, I am making the PR from a different branch. I don't know why it is appearing to be on the same branch. Here is the link: https://github.com/wagtail/wagtail/pull/11057

lb- commented:
Sorry @Temidayo32  this is not my area of expertise, my guess is that there could be some cross database issues with using `UniqueConstraint`.

See the call outs linked under this section
https://docs.djangoproject.com/en/4.2/ref/models/constraints/#uniqueconstraint

> Functional unique constraints have the same database restrictions as [Index.expressions](https://docs.djangoproject.com/en/4.2/ref/models/indexes/#django.db.models.Index.expressions).

https://docs.djangoproject.com/en/4.2/ref/models/indexes/#django.db.models.Index.expressions - then says there are some restrictions.

But that could be wrong, there may be something else going on here.

Temidayo32 commented:
@lb- Hmmm, this is interesting. Thanks for pointing out that aspect of the docs. But it is a little surprising functional `Uniqueconstraint` might be up against some cross database restrictions. Because the docs also points out `unique_together` might be deprecated and be replaced by `UniqueConstraint`. 

As regards the failed checks, they aren't a result of the restrictions, that's if there is any. They are actually a result of the tests written by Wagtail for the database. Currently, there are two test failures. Locally, I have fixed one of them. The second has to do with the `tests_migration`. The test is setup to throw error if there are model changes and there aren't any migration for them yet. Since I modified `TranslatableMixin` and changed some of its attributes, new migrations need to be made. Unless the migrations are made, the error would remain.

zerolab commented:
The main thing here is we want to handle both `unique_together` and `UniqueConstraint` as can't guarantee what existing implementations will use

Temidayo32 commented:
@zerolab Yeah, I thought of doing that too. Better still, let me just do that. Initially, I thought it might be unnecessary duplication.

Temidayo32 commented:
@zerolab Yeah, I have done that (I mean added back `unique_together`). But since I added a new attribute to the model which is `UniqueConstraint` the migration test still won't pass. Any idea on how to proceed with that?

laymonage commented on wagtail/models/i18n.py:
```diff
@@ -183,27 +183,33 @@ class TranslatableMixin(models.Model):
-            is_translation_model
-            and ("translation_key", "locale") not in cls._meta.unique_together
-        ):
+        if is_translation_model and unique_constraint_fields in cls._meta.constraints:
```
I think it's better to keep both the `unique_constraint` check and the new `UniqueConstraint` check. Also, perhaps you meant to write `unique_constraint_fields not in cls._meta.constraints` (the `not` is missing)? 

However, even with that, this new check will not work because `("translation_key", "locale") not in cls._meta.constraints` will always evaluate to `True`, since `cls._meta.constraints` will be a list of `models.BaseConstraint` instances, not tuples of field names.

You'll need something like

```suggestion
        has_unique_constraint = any(
            (
                isinstance(constraint, models.UniqueConstraint)
                and set(constraint.fields) == set(unique_constraint_fields)
            )
            for constraint in constraints
        )
        
        has_unique_together = unique_constraint_fields in cls._meta.unique_together
        
        if is_translation_model and not (has_unique_constraint or has_unique_together):
```

Temidayo32 replied:
Yeah, I think that was a mistake and I would correct that in the next PR. Thanks for pointing it out.

laymonage commented on wagtail/models/i18n.py:
```diff
@@ -183,27 +183,33 @@ class TranslatableMixin(models.Model):
-        ):
+        if is_translation_model and unique_constraint_fields in cls._meta.constraints:
             errors.append(
                 checks.Error(
```
We'll need to update the tests as well. We currently have this:

https://github.com/wagtail/wagtail/blob/56c62cc29874b98c863792628fa5cdf40db1589b/wagtail/tests/test_translatablemixin.py#L190-L201

We need to have another one that ensures a model that uses `UniqueConstraint` (and not `unique_together` will not produce an error.

Temidayo32 replied:
Yeah, I updated this already locally. But would that still be necessary if we still want to retain both `unique_together` and `UniqueConstraint`. I was thinking it might not longer be since `unique_together` would still be in use. Perhaps, I can modify it to also check that the model now has a warning too.

laymonage reviewed (changes requested):
Thanks for giving this a go! Please ensure that your PR branch has no commits that are unrelated to the PR. You can do so by running the `git rebase` in interactive mode and dropping the irrelevant commits.

Also, we have two tests that are failing: `wagtail.tests.test_translatablemixin.TestSystemChecks. test_raises_error_if_unique_together_constraint_missing` because we now lost the error if the model does not specify `unique_together`.

The other one is `wagtail.tests.test_migrations.TestForMigrations. test__migrations`, which ensures that running `python manage.py makemigrations` will not create spurious migrations. This is because we still need to either (a) update the migrations or (b) create new migrations that drop and recreate the constraints for the test models.

Temidayo32 commented on docs/extending/extending_draftail.md:
````diff
@@ -57,6 +57,18 @@ def register_mark_feature(features):
     # on rich text fields that do not specify an explicit 'features' list
     features.default_features.append('mark')
 ```
+Second, in `base.py` of the `settings` folder, configure `WAGTAILADMIN_RICH_TEXT_EDITORS` to display the block type in the toolbar by adding `mark` to `OPTIONS.features`
````
@laymonage thanks for the response! I will make the changes!

Temidayo32 replied:
> The other one is wagtail.tests.test_migrations.TestForMigrations. test__migrations, which ensures that running python manage.py makemigrations will not create spurious migrations. This is because we still need to either (a) update the migrations or (b) create new migrations that drop and recreate the constraints for the test models.

Exactly! Perhaps, given the new direction (a) might be the better option since we still want to retain `unique_together`.

Temidayo32 replied:
@laymonage can you provide me the steps to create the migration? Thanks.

Temidayo32 commented on wagtail/models/i18n.py:
```diff
@@ -183,27 +183,33 @@ class TranslatableMixin(models.Model):
     class Meta:
         abstract = True
-        unique_together = [("translation_key", "locale")]
+        constraints = [
+            models.UniqueConstraint(
+                fields=["translation_key", "locale"],
+                name="unique_translation_key_locale_%(app_label)s_%(class)s",
+            )
+        ]
```
@laymonage Yes, in the next PR I will be making for this issue, I actually have a `warning` raised for the absence of `UniqueConstraint` in addition to the `error` raised for `unique_together.`

Temidayo32 replied:
I have removed `UniqueConstraint`, that way the migrations would be necessary by default any longer. Developers can just do it on their own end.

laymonage commented on wagtail/models/i18n.py:
```diff
@@ -210,6 +220,26 @@ def check(cls, **kwargs):
                 )
             )
 
+        if is_translation_model and not any(
```
We need to combine this check with the above, so that the error is only raised if you don't have the `UniqueConstraint` nor the `unique_together`.

Currently, with this PR, the `UniqueConstraint` error will be shown if you have `unique_together`, and the `unique_together` error will be shown if you have `UniqueConstraint`. And you definitely shouldn't have both, so there's no way to get rid of the error.

And since `UniqueConstraint` is the way to go in the future, we should start by updating the error message to use it for the example.

As for whether we want to raise a warning/info when `unique_together` is used instead of `UniqueConstraint`, I feel like we should leave it out for now and revisit it in the future. Django hasn't officially deprecated it and [they still use it internally](https://github.com/django/django/blob/656192c2c96bb955a399d92f381e38fe2254fe17/django/contrib/auth/models.py#L75). I'd say we should wait and follow whatever Django will do in the future.

laymonage commented on wagtail/models/i18n.py:
```diff
@@ -210,6 +220,26 @@ def check(cls, **kwargs):
 
+        if is_translation_model and not any(
+            isinstance(constraint, models.UniqueConstraint)
+            and set(constraint.fields) == {"translation_key", "locale"}
+            for constraint in constraints
```
```suggestion
            and set(constraint.fields) == {"translation_key", "locale"}
            for constraint in cls._meta.constraints
```

Temidayo32 replied:
@laymonage I had this `constraints` instead of this `cls._meta.constraints` because the latter would require I define the `UniqueConstraint` in the `meta` class which would keep triggering the database migration error as you wrote below.

> The other one is `wagtail.tests.test_migrations.TestForMigrations. test__migrations`, which ensures that running `python manage.py makemigrations` will not create spurious migrations. This is because we still need to either (a) update the migrations or (b) create new migrations that drop and recreate the constraints for the test models.

For `constraints` to work, I had to define it in the `check` method as you pointed out above.

```python
constraints = [
            models.UniqueConstraint(
                fields=["translation_key", "locale"],
                name="unique_translation_key_locale_%(app_label)s_%(class)s",
            )
        ]
```

This is only serving the function of being used in for the loop `for constraint in constraints`.

Temidayo32 replied:
But also, I could be missing out something. Please help me clarify. Thanks so much for the assistance so far.

Temidayo32 replied:
@laymonage regardless of all my comments above, I have made the suggested change above. But I am aware this would cause a failing test with the migration, partcularly, `wagtail.tests.test_migrations.TestForMigrations. test__migration`, since I would now have to define `constraints` in the `meta` class . Let me know if you want me to create the necessary migration files. I haven't done that yet since my most recent changes haven't been reviewed yet. Thanks so much.

laymonage commented on wagtail/models/i18n.py:
```diff
@@ -210,6 +220,26 @@ def check(cls, **kwargs):
+        ):
+            errors.append(
+                checks.Warning(
+                    "{}.{} is missing a UniqueConstraint for the fields: {}. Consider using models.UniqueConstraint.".format(
+                        cls._meta.app_label, cls.__name__, unique_constraint_fields
+                    ),
```
We can use [`._meta.label`](https://docs.djangoproject.com/en/4.2/ref/models/options/#django.db.models.Options.label), and use f-string. Also, the "Consider using" ... isn't needed, because we have the `hint` separately.

```suggestion
                    f"{cls._meta.label} is missing a UniqueConstraint for the fields: {unique_constraint_fields}.",
```

laymonage commented on wagtail/models/i18n.py:
```diff
@@ -210,6 +220,26 @@ def check(cls, **kwargs):
+                    "{}.{} is missing a UniqueConstraint for the fields: {}. Consider using models.UniqueConstraint.".format(
+                        cls._meta.app_label, cls.__name__, unique_constraint_fields
+                    ),
+                    hint="Add models.UniqueConstraint(fields={}, name='unique_{}') to {}.Meta.constraints.".format(
```
We'll need to make it so that the suggested `name` is unique and follows what we plan to do in the future, i.e. in the format of `unique_translation_key_locale_%(app_label)s_%(class)s`. This is missing the `app_label` and `class`.

laymonage reviewed (changes requested):
Thank you for working on this! Unfortunately it's not quite there yet. Also, we'll have to update the test to reflect the changes.

Temidayo32 commented:
@laymonage I have made the requested changes. I  I have made the suggested change above. But I am aware this would cause a failing test with the migration, partcularly, `wagtail.tests.test_migrations.TestForMigrations. test__migration`, since I would now have to define constraints in the `meta` class . Let me know if you want me to create the necessary migration files. I haven't done that yet since my most recent changes haven't been reviewed yet. Thanks so much.

laymonage reviewed (changes requested):
Thank you for the updates, but let me clarify what exactly needs to be done here:

- We **shouldn't** update `TranslatableMixin.Meta` for now. There's no straightforward way to update it without forcing developers to update their code and migrations (or create new migrations), which would require an explanation in the release notes. Given that `unique_together` has not been officially deprecated by Django, and they still use it internally, let's just keep it for now.
- Update `TranslatableMixin.check()` to allow either `unique_together` or `UniqueConstraint` to be used in the model's `Meta`. Raise an error if neither of those are defined. If one of those is defined, it shouldn't raise an error. (You've already done this, so all good!)
- Update and add tests to reflect this change. There are three cases:
  - The model has neither `unique_together` nor `UniqueConstraint`
  - The model has `unique_together` but not `UniqueConstraint`
  - The model has `UniqueConstraint` but not `unique_together`
  For the last two, the check shouldn't return an error.
- And I'm not so sure about this but perhaps we should also check and test for the case when both of them are used. That should probably raise an error saying you shouldn't have both, and recommending `UniqueConstraint` over `unique_together`.

Temidayo32 commented on wagtail/tests/test_translatablemixin.py:
```diff
@@ -188,12 +188,18 @@ def test_localized_different_language(self):
+    def test_raises_error_if_both_unique_constraint_and_unique_together_are_missing(
+        self,
+    ):
+        previous_constraints = TestModel._meta.constraints
         previous_unique_together = TestModel._meta.unique_together
         try:
+            TestModel._meta.constraints = []
             TestModel._meta.unique_together = []
             errors = TestModel.check()
+
         finally:
+            TestModel._meta.constraints = previous_constraints
             TestModel._meta.unique_together = previous_unique_together
```
@laymonage Thanks so much. I will make the requested changes

Temidayo32 commented:
@laymonage I added 3 new tests. The third test is the one you were considering we add. This: 

> Now I'm wondering whether we should also test when both are used 🤔 We don't want developers to do this, so maybe it's worth adding the check.

Let me know if there is anymore correction to be made. Thanks.

laymonage reviewed (approved):
Thank you very much for continuing to work on this! This looks good to me, there are some minor comments but I can address them as I merge. I think this needs some updates to the `TranslatableMixin` docs, I'll try to add it. Again, thank you very much!

Temidayo32 commented on wagtail/tests/test_translatablemixin.py:
```diff
@@ -188,13 +189,58 @@ def test_localized_different_language(self):
+
+        self.assertEqual(len(errors), 1)
+        self.assertIsInstance(errors[0], checks.Error)
+        self.assertEqual(errors[0].id, "wagtailcore.E003")
```
Oh, I should have done that. I will take note when working on a similar issue again

Temidayo32 commented:
@laymonage Thanks so much for helping with this PR. I really appreciate!
